File: 1569-max-dot-product-of-two-subsequences/1569-max-dot-product-of-two-subsequences.java
import static java.lang.Math.*;
class Solution {
    Integer[][][] dp;
    int n, m;
    int[] a, b;
    final int inf = (int)(1e9);
    public int maxDotProduct(int[] nums1, int[] nums2) {
        n = nums1.length;
        m = nums2.length;
        this.a = nums1;
        this.b = nums2;
        dp = new Integer[n][m][2];
        return recur(0, 0, 0);
    }
    public int recur(int i, int j, int k) {
        if(i == n || j == m) {
            if(k == 1) return 0;
            return -inf;
        }
        if(dp[i][j][k] != null) return dp[i][j][k];
        int ans = -inf;
        ans = max(ans, a[i] * b[j] + recur(i + 1, j + 1, 1));
        ans = max(ans, recur(i + 1, j, k));
        ans = max(ans, recur(i, j + 1, k));
        return dp[i][j][k] = ans;
    }
}

File: 1569-max-dot-product-of-two-subsequences/1569-max-dot-product-of-two-subsequences.py
class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        n = len(nums1)
        m = len(nums2)
        inf = 1e9

        dp = [[[None] * (2) for _ in range(m)] for _ in range(n)]

        def recur(i, j, k):
            if i == n or j == m:
                if k == 1:
                    return 0
                return -inf

            if dp[i][j][k] != None:
                return dp[i][j][k]

            ans = -inf
            ans = max(ans, nums1[i] * nums2[j] + recur(i + 1, j + 1, 1))
            ans = max(ans, max(recur(i + 1, j ,k), recur(i, j + 1, k)))

            dp[i][j][k] = ans

            return dp[i][j][k]
        
        return recur(0, 0, 0)